File: ui/lib/encryption.ts
import { createCipheriv, createDecipheriv, randomBytes } from "crypto";

function getKey(): Buffer {
  const raw = process.env.ENCRYPTION_KEY;
  if (!raw) throw new Error("ENCRYPTION_KEY env var is not set");
  const padded = raw.replace(/-/g, "+").replace(/_/g, "/");
  const buf = Buffer.from(padded, "base64");
  if (![16, 24, 32].includes(buf.length))
    throw new Error(`ENCRYPTION_KEY must decode to 16/24/32 bytes, got ${buf.length}`);
  return buf;
}

export function encrypt(plaintext: string): string {
  const key = getKey();
  const nonce = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  const ct = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  const combined = Buffer.concat([nonce, ct, tag]);
  return combined.toString("base64url");
}

export function decrypt(ciphertext: string): string {
  const key = getKey();
  const raw = Buffer.from(ciphertext, "base64url");
  const nonce = raw.subarray(0, 12);
  const tag = raw.subarray(raw.length - 16);
  const ct = raw.subarray(12, raw.length - 16);
  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ct), decipher.final()]).toString("utf8");
}

File: src/services/encryption.py
"""AES-256-GCM symmetric encryption for API keys stored in Supabase.

Generate a key once and set it in both .env (Python) and .env.local (Next.js):
    python -c "import secrets,base64; print(base64.urlsafe_b64encode(secrets.token_bytes(32)).decode())"

Stored format: base64url(nonce[12] + ciphertext + tag[16])
The Node.js counterpart in ui/lib/encryption.ts uses the identical scheme so
keys encrypted by Next.js can be decrypted here and vice versa.
"""
from __future__ import annotations

import base64
import os
import secrets as _secrets

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _key() -> bytes:
    raw = os.getenv("ENCRYPTION_KEY", "")
    if not raw:
        raise RuntimeError(
            "ENCRYPTION_KEY env var is not set. "
            "Generate one with: python -c \"import secrets,base64; "
            "print(base64.urlsafe_b64encode(secrets.token_bytes(32)).decode())\""
        )
    padded = raw + "==" * (4 - len(raw) % 4 if len(raw) % 4 else 0)
    try:
        key = base64.urlsafe_b64decode(padded)
    except Exception as e:
        raise RuntimeError(f"ENCRYPTION_KEY is not valid base64url: {e}") from e
    if len(key) not in (16, 24, 32):
        raise RuntimeError(f"ENCRYPTION_KEY must decode to 16/24/32 bytes, got {len(key)}")
    return key


def encrypt(plaintext: str) -> str:
    """Encrypt plaintext string → base64url-encoded ciphertext."""
    aesgcm = AESGCM(_key())
    nonce  = _secrets.token_bytes(12)
    ct     = aesgcm.encrypt(nonce, plaintext.encode("utf-8"), None)
    return base64.urlsafe_b64encode(nonce + ct).decode("ascii")


def decrypt(ciphertext: str) -> str:
    """Decrypt base64url-encoded ciphertext → plaintext string."""
    aesgcm = AESGCM(_key())
    raw    = base64.urlsafe_b64decode(ciphertext + "==" * (4 - len(ciphertext) % 4 if len(ciphertext) % 4 else 0))
    nonce, ct = raw[:12], raw[12:]
    return aesgcm.decrypt(nonce, ct, None).decode("utf-8")
